R/export_tibble.R

```diff
@@ -51,53 +52,93 @@ check_var_type <- function(GADSdat) {
+     identical(varClass, "character")) {
+    attr_list[["na_values"]] <- as.character(attr_list[["na_values"]])
     if(nrow(value_label_df) > 0) {
-      out[["labels"]] <- as.character(out[["labels"]])
-      names(out[["labels"]]) <- value_label_df[, "valLabel"]
+      attr_list[["labels"]] <- as.character(attr_list[["labels"]])
+      names(attr_list[["labels"]]) <- value_label_df[, "valLabel"]
     }
-    #if(identical(labeled, "yes")) out[["class"]] <- c("haven_labelled")
+    #if(identical(labeled, "yes")) attr_list[["class"]] <- c("haven_labelled")
   }
 
-  #if(any(label_df$varName == "TESTUNG")) browser()
-  out
+  attr_list
 }
 
+# missing tag conversion to suit SPSS style (maximum 3 discrete missing tags or range of missing tags)
+add_miss_tags <- function(varName, attr_list, label_df, raw_dat) {
+  miss_values <- label_df[which(label_df$missings == "miss"), "value"]
+
+  if(length(miss_values) > 0 && length(miss_values) <= 3)  {
+    attr_list[["na_values"]] <- miss_values
+  }
+  if(length(miss_values) > 3) {
+    full_range <- range(miss_values)
+    # check variable class
+    if(is.character(raw_dat)) {
+      stop("Missing tag conversion for variable '", varName,
+           "' to SPSS conventions is not possible, as too many missings are declared for a character variable (maximum 3). Adjust missing tags to export to tibble or write to SPSS.")
+    }
+    # check other labels
+    non_miss_labels <- label_df[is.na(label_df$missings) | label_df$missings == "valid", "value"]
+    accidental_miss_tags_labels <- non_miss_labels[!is.na(non_miss_labels) &
+                                                     non_miss_labels >= full_range[1] & non_miss_labels <= full_range[2]]
+    if(length(accidental_miss_tags_labels) > 0) {
+      stop("Missing tag conversion for variable '", varName,
+           "' to SPSS conventions is not possible, as the new missing range (",
+           full_range[1], " to ", full_range[2],
+           ") would include the following labeled values not tagged as missing: ",
+           paste(accidental_miss_tags_labels, collapse = ", "),
+           ". Adjust missing tags to export to tibble or write to SPSS.")
+    }
+
+    # check actual data
+    other_raw_dat <- raw_dat[!raw_dat %in% miss_values]
+    accidental_miss_tags_data <- other_raw_dat[!is.na(other_raw_dat) &
+                                            other_raw_dat >= full_range[1] & other_raw_dat <= full_range[2]]
+    if(length(accidental_miss_tags_data) > 0) {
+      stop("Missing tag conversion for variable '", varName,
+           "' to SPSS conventions is not possible, as the new missing range (",
+           full_range[1], " to ", full_range[2],
+           ") would include the following values not tagged as missing in the data: ",
+              paste(accidental_miss_tags_data, collapse = ", "),
+           ". Adjust missing tags to export to tibble or write to SPSS.")
+    }
+
+    attr_list[["na_range"]] <- full_range

```

Comment:
LGTM. Nit: The messages could be adapted if my proposal should be used.

However (minor nit!), the messages are very parallel as only the body differs. Maybe, these elements could be refactored in the future to something like this:

 ```r
missing_error_message <- function(body) {
   stop("Missing tag conversion for variable '", varName, "' to SPSS conventions is not possible.",
        body,
        "Adjust missing tags to export to tibble or write to SPSS.")
}
```

Reply:
I do agree that such a refactoring could be worthwile, but for now I would leave it as is (works as intended and refactoring would be more work 😛 ). But thanks for the comment regardless, will keep that in mind for future, similar cases!

Reply:
I totally agree!

If you like, you can turn this comment into an issue, I guess (with "Reference in new issue" or so?).